src/main/java/com/playmonumenta/networkrelay/RabbitMQManager.java

```diff
@@ -138,11 +202,23 @@ protected RabbitMQManager(Plugin plugin, String shardName, String rabbitURI) thr
 	protected void stop() {
 		mShutdown = true;
 		if (mConsumerAlive) {
+			try {
+				mHeartbeatRunnable.cancel();
+			} catch (Exception ex) {
+				mLogger.warning("Failed to cancel heartbeat runnable: " + ex.getMessage());
+			}
 			try {
 				mChannel.basicCancel(CONSUMER_TAG);
 			} catch (Exception ex) {
 				mLogger.warning("Failed to cancel rabbit consumer: " + ex.getMessage());
 			}
+			try {
+				JsonObject data = new JsonObject();
+				data.addProperty("online", false);
+				sendNetworkMessage("*", NetworkRelayAPI.HEARTBEAT_CHANNEL, data);
+			} catch (Exception ex) {
+				mLogger.warning("Failed to send shutdown heartbeat: " + ex.getMessage());
+			}
```

Comment:
I don't see how this can work - pretty sure sendNetworkMessage is async, and you close the connection immediately after. There's no way the data is going to be sent in time. You need some kind of blocking API call here that prevents stop from proceeding until the message is sent.

Reply:
Ah, found something - https://www.rabbitmq.com/tutorials/tutorial-seven-java.html

```java
mChannel.waitForConfirmsOrDie(5000); // Block up to 5s waiting for messages to be confirmed
```

Reply:
Waiting for confirmation requires calling one of these methods first, none of which we've used:
https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/Channel.html#addConfirmListener(com.rabbitmq.client.ConfirmCallback,com.rabbitmq.client.ConfirmCallback)
https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/Channel.html#addConfirmListener(com.rabbitmq.client.ConfirmListener)
https://rabbitmq.github.io/rabbitmq-java-client/api/current/com/rabbitmq/client/Channel.html#confirmSelect()